File: src/main/java/chat/teco/tecochat/chat/domain/chat/SettingMessage.java
package chat.teco.tecochat.chat.domain.chat;

import static chat.teco.tecochat.chat.domain.chat.Role.SYSTEM;

import chat.teco.tecochat.member.domain.Course;
import java.util.EnumMap;
import java.util.Map;

public enum SettingMessage implements Message {

    // (참고) 토큰 수는 31 나옴
    BACK_END_SETTING("You are a helpful backend developer assistant."),
    FRONT_END_SETTING("You are a helpful frontend developer assistant."),
    ANDROID_SETTING("You are a helpful android developer assistant."),
    ;

    private static final String DEFAULT_LANGUAGE_SETTING = "If there is no request, please reply in Korean.";
    private static final Map<Course, SettingMessage> byCourseMap;

    static {
        byCourseMap = new EnumMap<>(Course.class);
        byCourseMap.put(Course.BACKEND, BACK_END_SETTING);
        byCourseMap.put(Course.FRONTEND, FRONT_END_SETTING);
        byCourseMap.put(Course.ANDROID, ANDROID_SETTING);
    }

    private final String content;

    SettingMessage(String message) {
        this.content = message + DEFAULT_LANGUAGE_SETTING;
    }

    public static SettingMessage byCourse(Course course) {
        return byCourseMap.get(course);
    }

    public String message() {
        return content;
    }

    @Override
    public String roleName() {
        return SYSTEM.roleName();
    }

    @Override
    public String content() {
        return content;
    }
}

File: src/main/kotlin/chat/teco/tecochat/domain/chat/SettingMessage.kt
package chat.teco.tecochat.domain.chat

import chat.teco.tecochat.domain.member.Course
import java.util.EnumMap

private const val DEFAULT_LANGUAGE_SETTING = "If there is no request, please reply in Korean."

enum class SettingMessage(message: String) : Message {
    // (참고) 토큰 수는 31 나옴
    BACK_END_SETTING("You are a helpful backend developer assistant."),
    FRONT_END_SETTING("You are a helpful frontend developer assistant."),
    ANDROID_SETTING("You are a helpful android developer assistant.");

    private val content: String

    init {
        content = message + DEFAULT_LANGUAGE_SETTING
    }

    fun message(): String {
        return content
    }

    override fun roleName(): String {
        return Role.SYSTEM.roleName
    }

    override fun content(): String {
        return content
    }

    companion object {
        private val byCourseMap: MutableMap<Course, SettingMessage>

        init {
            byCourseMap = EnumMap(Course::class.java)
            byCourseMap[Course.BACKEND] = BACK_END_SETTING
            byCourseMap[Course.FRONTEND] = FRONT_END_SETTING
            byCourseMap[Course.ANDROID] = ANDROID_SETTING
        }

        @JvmStatic
        fun byCourse(course: Course): SettingMessage {
            return byCourseMap[course] ?: BACK_END_SETTING
        }
    }
}
